### PR TITLE
eID-Service implementation

### PR DESCRIPTION
**Pull request by BenjaminKeller**
As required by #19 and #22.

@nils-wisiol You've requested to mark some lines in get_tc_token_url as important security features. How can I achieve that? By leaving a comment?

### CODE REVIEW

File: api/entrypoint.sh [hunk 1]
```diff
@@ -8,6 +8,8 @@ python manage.py collectstatic --clear --noinput || exit 1
 
 # migrate database
 python manage.py migrate || exit 1
+python manage.py makemigrations eid_server || exit 1
```
Comment (nils-wisiol) [inline api/entrypoint.sh]: `makemigrations` should only be called during development time. That is, please call makemigrations on your dev machine, have the migrations generated and add them to this repository.

File: api/entrypoint.sh [hunk 2]
```diff
@@ -8,6 +8,8 @@ python manage.py collectstatic --clear --noinput || exit 1
...
+python manage.py migrate eid_server || exit 1
```
Comment (nils-wisiol) [inline api/entrypoint.sh]: After moving the migrations to the repository, this line becomes obsolete as migrations will be applied by the previous migrate command.

File: test/e2e/spec/tcToken_spec.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+				'Host': process.env.BOILERPLATE_DOMAIN,
+				'Connection': 'close',
+			}
+		  }
+		});
+		
+		frisby.addExpectHandler('tc token', function(response) {
+			let xml = response._body.replace(/\n/g, "").replace(/\r/g, "").replace(/ /g, "");
+
+            expect(xml).toEqual(jasmine.stringMatching(/^<TCTokenType><ServerAddress>.*<\/Binding><\/TCTokenType>$/));
```
Comment (nils-wisiol) [inline test/e2e/spec/tcToken_spec.js]: please fix indentation (also, throughout the file)

**Review by nils-wisiol — CHANGES_REQUESTED** [review]
looks good, just a few changes requested :+1:

**Review by Armagetron — CHANGES_REQUESTED** [review]
Please move the TCToken to the eID Service since the eID Server is not responsible of creating the token

**Comment by BenjaminKeller** [comment]
@nils-wisiol Thank you for your feedback and the hint with the migrations. I corrected everything.
@Armagetron I fix it.

**Comment by BenjaminKeller** [comment]
Renamed eid_server to eid_service.

**Review by Armagetron — APPROVED** [review]

File: test/e2e/spec/tcTokenUrl_spec.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		  request: {
+			headers: {
+				'Host': process.env.BOILERPLATE_DOMAIN,
+				'Connection': 'close',
+			}
+		  }
+		});
+		
+		frisby.addExpectHandler('correct redirect', function(response) {
+			let header = response.headers._headers;
```
Comment (nils-wisiol) [inline test/e2e/spec/tcTokenUrl_spec.js]: indentation is not in accordance with #30, please update accordingly

File: test/e2e/spec/tcToken_spec.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
+describe("The tcToken generator", function() {
+
+  var frisby = require('frisby');
+  var URL = 'https://' + process.env.BOILERPLATE_DOMAIN + '/api/getTcToken';
+
+  beforeAll(function() {
+    frisby.globalSetup({ // globalSetup is for ALL requests
+      request: {
+      headers: {
+        'Host': process.env.BOILERPLATE_DOMAIN,
```
Comment (nils-wisiol) [inline test/e2e/spec/tcToken_spec.js]: indentation is not in accordance with #30

File: api/boilerplate/views.py [hunk 6]
```diff
@@ -1,7 +1,47 @@
...
+        except ValueError:
+            return HttpResponseBadRequest()
+        except TypeError:
+            return HttpResponseBadRequest()
+        try:
+            authenticationRequest = AuthenticationRequest.objects.get(tcTokenId=tcTokenId)
+        except AuthenticationRequest.DoesNotExist:
+            return HttpResponseBadRequest()
+        context = {
+            'serverAddress': 'https://...',
```
Comment (nils-wisiol) [inline api/boilerplate/views.py]: Is `...` actually sent to the client? Is that expected behavior?

**Review by nils-wisiol — COMMENTED** [review]
I have some confusion about the response of `getTcToken`, which seems to include `...` has a hostname. This is not covered by the spec, please clarify

File: api/boilerplate/views.py (hunk 6 above)
Comment (BenjaminKeller) [reply api/boilerplate/views.py]: This is a stub and must be filled by following pull requests implementing the eID-Server.

**Comment by BenjaminKeller** [comment]
@nils-wisiol As already mentioned `...` is a stub and must be filled by following pull requests implementing the eID-Server. Sorry for the intendation mistake, I rechecked all files.

**Comment by Armagetron** [comment]
The eID will be most probably running at BOILERPLATE_DOMAIN/eID

**Comment by BenjaminKeller** [comment]
OK, I adapt this and make use of the UseId-Call.

**Comment by BenjaminKeller** [comment]
I moved the implementation into its own folder and created a python module for it. I added comments where the implementation of the eID-Server is missing.

File: api/boilerplate/views.py [hunk 7]
```diff
@@ -1,7 +1,47 @@
+from uuid import UUID
 from django.contrib.auth.models import User, Group
 from rest_framework import viewsets
 from .serializers import UserSerializer, GroupSerializer
+from django.shortcuts import redirect, render
+from django.http import HttpResponseBadRequest
+from eid_service.models import AuthenticationRequest
 
+def get_tc_token_url(request):
```
Comment (nils-wisiol) [inline api/boilerplate/views.py]: with the updated `urls.py`, isn't that code here dead?

**Review by nils-wisiol — CHANGES_REQUESTED** [review]
- I believe there is some dead code added to the `boilerplate` module.
- tests are broken as of now

**Comment by nils-wisiol** [comment]
Could you please reiterate the difference of eid service and eid server? I didn't quite understand it yet. (sorry!)

File: api/boilerplate/views.py (hunk 7 above)
Comment (BenjaminKeller) [reply api/boilerplate/views.py]: Sorry for that. I removed it.

**Comment by nils-wisiol** [comment]
tests are still broken

**Comment by Armagetron** [comment]
I seriously hoped that the increase of the requests would help to solve this issue with the failing tests. I have no idea why they fail.

**Comment by BenjaminKeller** [comment]
@zervnet As already discussed on friday we have problems with too many connections to nginx/django. It has to be a limitation on the technical side because the tests fail randomly. The following error message is provided:
```
api_1     | [pid: 157|app: 0|req: 1/109] 10.0.0.2 () {40 vars in 584 bytes} [Tue Aug 22 06:31:33 2017] GET /api/eIdService/getTcTokenUrl?host=127.0.0.1 => generated 0 bytes in 361 msecs (HTTP/1.1 302) 4 headers in 257 bytes (1 switches on core 0)
api_1     | Tue Aug 22 06:31:33 2017 - uwsgi_response_write_body_do(): Broken pipe [core/writer.c line 331] during GET /api/eIdService/getTcTokenUrl?protocol=eid (10.0.0.2)
api_1     | OSError: write error
```

The following configurations didn't solve the problem:
```
www/conf/nginx.conf:

 events {
-    worker_connections  1024;
+    worker_connections  2048;
 }

www/conf/sites-available/20-static-and-api.conf.var:

   location /api/ {
     # max .5r/s to the API per IP, but 10 at once is okay. This limit may need to be increased once
     # client applications become more powerful
-    limit_req zone=perip-api burst=10 nodelay;
+    limit_req zone=perip-api burst=1000 nodelay;
     expires epoch;
     etag off;
     add_header Strict-Transport-Security "max-age=31536000";
```

**Comment by nils-wisiol** [comment]
There are indeed two tests failing, however the listing of *which* tests fail seems to be broken in frisby. I am looking into this.

**Comment by nils-wisiol** [comment]
I came to the conclusion that frisby is broken. With working tests, everything is fine. But with broken tests, it reports successful tests to be broken and vice-versa.

I propose migrating to an alternative testing framework, possibly [chakram](http://dareid.github.io/chakram/). Their examples are extensive and it seems to be better maintained in general, too.

Please also see #54

**Comment by BenjaminKeller** [comment]
@nils-wisiol Can we merge now? Or do I have to rewrite history?

**Comment by nils-wisiol** [comment]
yes, please rewrite history :+1:

**Comment by BenjaminKeller** [comment]
@nils-wisiol Done.

**Comment by nils-wisiol** [comment]
Fair enough. Next time, please use rebase instead of actually deleting the history to keep `git blame` instact.

**Review by nils-wisiol — APPROVED** [review]

**Pull request merged**